### PR TITLE
Improve spotlessApply discovery for multi gradle project & add support for vscode-gradle@3.7.0

### PR DESCRIPTION
**Pull request by forestaa**
This PR fixes the problem that the `spotlessApply` task cannot be discovered under multi Gradle project structure.

Suppose the multi Gradle project below. `app` and `lib` are subprojects and `app` project have the spotless plugin.
```
root
├── app
│   ├── build.gradle  <-- Have the spotless plugin
│   └── src/main/java/com/example/Main.java
├── lib
│   └── build.gradle
├── build.gradle
└── settings.gradle
```

In this case, the root project doesn't have the `spotlessApply` task, but `./gradlew spotlessApply` will execute the `spotlessApply` in any subprojects.
This time, `./graldew spotlessApply` will execute `./gradlew app:spotlessApply`.
ref. https://docs.gradle.org/current/userguide/intro_multi_project_builds.html#sec:executing_tasks_by_name

<details>
<summary> ./gradlew tasks --all </summary>
<blockquote>
$ ./gradlew tasks --all  

> Task :tasks

------------------------------------------------------------
Tasks runnable from root project 'basic-multiproject'
------------------------------------------------------------

Application tasks
-----------------
app:run - Runs this project as a JVM application

Build tasks
-----------
app:assemble - Assembles the outputs of this project.
app:build - Assembles and tests this project.
app:buildDependents - Assembles and tests this project and all projects that depend on it.
app:buildNeeded - Assembles and tests this project and all projects it depends on.
app:classes - Assembles main classes.
app:clean - Deletes the build directory.
app:jar - Assembles a jar archive containing the main classes.
app:testClasses - Assembles test classes.

Build Setup tasks
-----------------
init - Initializes a new Gradle build.
wrapper - Generates Gradle wrapper files.

Distribution tasks
------------------
app:assembleDist - Assembles the main distributions
app:distTar - Bundles the project as a distribution.
app:distZip - Bundles the project as a distribution.
app:installDist - Installs the project as a distribution as-is.

Documentation tasks
-------------------
app:javadoc - Generates Javadoc API documentation for the main source code.

Help tasks
----------
buildEnvironment - Displays all buildscript dependencies declared in root project 'basic-multiproject'.
app:buildEnvironment - Displays all buildscript dependencies declared in project ':app'.
lib:buildEnvironment - Displays all buildscript dependencies declared in project ':lib'.
dependencies - Displays all dependencies declared in root project 'basic-multiproject'.
app:dependencies - Displays all dependencies declared in project ':app'.
lib:dependencies - Displays all dependencies declared in project ':lib'.
dependencyInsight - Displays the insight into a specific dependency in root project 'basic-multiproject'.
app:dependencyInsight - Displays the insight into a specific dependency in project ':app'.
lib:dependencyInsight - Displays the insight into a specific dependency in project ':lib'.
help - Displays a help message.
app:help - Displays a help message.
lib:help - Displays a help message.
javaToolchains - Displays the detected java toolchains.
app:javaToolchains - Displays the detected java toolchains.
lib:javaToolchains - Displays the detected java toolchains.
outgoingVariants - Displays the outgoing variants of root project 'basic-multiproject'.
app:outgoingVariants - Displays the outgoing variants of project ':app'.
lib:outgoingVariants - Displays the outgoing variants of project ':lib'.
projects - Displays the sub-projects of root project 'basic-multiproject'.
app:projects - Displays the sub-projects of project ':app'.
lib:projects - Displays the sub-projects of project ':lib'.
properties - Displays the properties of root project 'basic-multiproject'.
app:properties - Displays the properties of project ':app'.
lib:properties - Displays the properties of project ':lib'.
tasks - Displays the tasks runnable from root project 'basic-multiproject' (some of the displayed tasks may belong to subprojects).
app:tasks - Displays the tasks runnable from project ':app'.
lib:tasks - Displays the tasks runnable from project ':lib'.

Verification tasks
------------------
app:check - Runs all checks.
app:spotlessApply - Applies code formatting steps to sourcecode in-place.
app:spotlessCheck - Checks that sourcecode satisfies formatting steps.
app:spotlessDiagnose
app:test - Runs the unit tests.

Other tasks
-----------
app:compileJava - Compiles main Java source.
app:compileTestJava - Compiles test Java source.
components - Displays the components produced by root project 'basic-multiproject'. [deprecated]
app:components - Displays the components produced by project ':app'. [deprecated]
lib:components - Displays the components produced by project ':lib'. [deprecated]
dependentComponents - Displays the dependent components of components in root project 'basic-multiproject'. [deprecated]
app:dependentComponents - Displays the dependent components of components in project ':app'. [deprecated]
lib:dependentComponents - Displays the dependent components of components in project ':lib'. [deprecated]
model - Displays the configuration model of root project 'basic-multiproject'. [deprecated]
app:model - Displays the configuration model of project ':app'. [deprecated]
lib:model - Displays the configuration model of project ':lib'. [deprecated]
prepareKotlinBuildScriptModel
app:processResources - Processes main resources.
app:processTestResources - Processes test resources.
spotlessInternalRegisterDependencies
app:spotlessJava
app:spotlessJavaApply
app:spotlessJavaCheck
app:spotlessJavaDiagnose
app:startScripts - Creates OS specific scripts to run the project as a JVM application.

BUILD SUCCESSFUL in 1s
1 actionable task: 1 executed
</blockquote>
</details>

The current implementation can discover the `spotlessApply` but cannot discover the `app:spotlessApply`.
So, this PR improves the discovery of `spotlessApply` task so that it discovers the `app:spotlessApply`.

### CODE REVIEW

**Comment by badsyntax** [comment]
Thank you for the PR! The change makes sense to me. I wonder if you'd be willing to write a test for this change? If not I can write a test for this.

**Comment by badsyntax** [comment]
Thanks for this. I have enabled the GitHub Actions but can only review the changes maybe tomorrow or in the near future.

**Comment by badsyntax** [comment]
Unfortunately the linting task is failing :( Gah that's annoying. I will have to sort this out. Sorry for this state of this repo. You can comment out the linting step in the actions workflow file in the meantine to get the pipelines to pass until i sort this out.

**Comment by forestaa** [comment]
This is my mistake, I forgot to run `npm run lint:fix`.
I ran lint and push the change.

I found that I need to add a build step for gradle-multi-project to `.github/workflows/main.yml` and added it.
Take a look when you have time.

**Comment by badsyntax** [comment]
Thanks again for this. The reason why the tests are now failing is because recently `vscode-gradle` was transferred to Microsoft but there's a problem with the publisher id. (See https://github.com/badsyntax/vscode-spotless-gradle/pull/242). I am speaking with the MS devs to try sort this out (see https://github.com/microsoft/vscode-gradle/issues/1007).

**Comment by forestaa** [comment]
I see. Thank you for your help!

I also fixed the problem that the path to multi gradle project is invalid.

**Comment by badsyntax** [comment]
Great. I have pushed some changes to master to fix all these issues. Can you update this branch with master?

**Comment by forestaa** [comment]
I got it. I merged upstream/master.

But the test fails due to time out. hmm...
I'll take a look later.

**Comment by badsyntax** [comment]
The reason for the timeout is because it's running the tests on the wrong fixture directory: https://github.com/forestaa/vscode-spotless-gradle/blob/8385d771b171e7ff6dd3f251f1295607dde0e3f3/src/test/runTest.ts#L17-L20

We will probably have to split up the tests into different directories. That will allow us to run the tests on different test-fixtures. This is the approach I took with `vscode-gradle`, see https://github.com/microsoft/vscode-gradle/blob/d288ef0c116023b229a1dbdef3566cac9470e7a7/extension/src/test/runTests.ts

Feel free to make that change, otherwise I am happy to do it.

**Comment by badsyntax** [comment]
Note for self: tests are failing with vscode-gradle@3.7.0 due to a change in how the extension is activated (it's not activated until you select the gradle view in the sidebar)

**Comment by forestaa** [comment]
Thank you for your advice. That helps me a lot.

I separate the test and hope the test will be fixed.

(I fixed my environment problem and finally can run tests locally, but the same issue you mentioned below happens to my environment...)
https://github.com/badsyntax/vscode-spotless-gradle/pull/241#issuecomment-924615058

**Comment by badsyntax** [comment]
@forestaa can you double-check you've installed the latest npm packages locally? maybe a good idea to delete node_modules and run `npm i` again.

**Comment by forestaa** [comment]
You are right, I forgot to install the latest npm package, sorry for my mistake...

**Comment by badsyntax** [comment]
Ok it's looking good. The final problem is now the timeouts in the tests. The reason for this is because the Gradle extension is not activated, due to a change in [vscode-gradle@3.7.0](https://github.com/microsoft/vscode-gradle/releases/tag/3.7.0). 

Before the tests run we need to select the Gradle view in the activitiy bar. That will then enable the Gradle extension. I am just searching for the appropriate vscode API to do this...

**Comment by badsyntax** [comment]
Can you add the following change to the [register method](https://github.com/badsyntax/vscode-spotless-gradle/blob/2a90236101b2384628e204e406f74972d73ce369/src/Spotless.ts#L22):

```diff
diff --git a/src/Spotless.ts b/src/Spotless.ts
index 1423edc..0edb1e5 100644
--- a/src/Spotless.ts
+++ b/src/Spotless.ts
@@ -28,6 +28,7 @@ export class Spotless {
           this.readyHandlers.forEach((handler) => handler(this.isReady));
         })
     );
+    this.gradleApi.getTaskProvider().provideTasks();
   }
 
   public dispose(): void {
```

That should trigger the vscode-gradle extension to get the list of gradle tasks which are required for this spotless extension to work in the tests.

**Comment by badsyntax** [comment]
Woop woop, tests are passing 🎉  Thank you again for your patience and for you contributions! I'd done a quick review and there's one or two improvements we can make, but we are close. I will do a proper review this evening.

**Comment by forestaa** [comment]
Thank you for your taking a lot of time for me too!
I'm waiting for your proper review!

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -46,6 +46,20 @@ jobs:
           build-root-directory: test-fixtures/gradle-project
           wrapper-directory: test-fixtures/gradle-project
           wrapper-cache-enabled: false
+      - name: Prepare Gradle
```
Comment (badsyntax) [inline .github/workflows/main.yml:49]:
```suggestion
      - name: Prepare Gradle (multi-project)
```

File: .github/workflows/main.yml [hunk 2]
```diff
@@ -46,6 +46,20 @@ jobs:
...
+        uses: eskatos/gradle-command-action@v1
+        with:
+          arguments: build -x spotlessCheck
+          build-root-directory: test-fixtures/gradle-multi-project
+          wrapper-directory: test-fixtures/gradle-multi-project
+          wrapper-cache-enabled: false
+      - name: Prepare Spotless
```
Comment (badsyntax) [inline .github/workflows/main.yml:56]:
```suggestion
      - name: Prepare Spotless (multi-project)
```

File: .github/workflows/main.yml [hunk 3]
```diff
@@ -66,6 +80,13 @@ jobs:
           build-root-directory: test-fixtures/gradle-project
           wrapper-directory: test-fixtures/gradle-project
           wrapper-cache-enabled: false
+      - name: Stop Gradle daemon
```
Comment (badsyntax) [inline .github/workflows/main.yml:83]:
```suggestion
      - name: Stop Gradle daemon (multi-project)
```

File: test-fixtures/gradle-multi-project/.vscode/settings.json [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+  "typescript.format.enable": false,
+  "javascript.validate.enable": false,
+  "javascript.format.enable": false,
+  "spotlessGradle.diagnostics.enable": false,
+  "spotlessGradle.format.enable": true,
+  "[java]": {
+    "editor.defaultFormatter": "richardwillis.vscode-spotless-gradle",
+    "spotlessGradle.diagnostics.enable": true
+  },
+  "[kotlinscript]": {
```
Comment (badsyntax) [inline test-fixtures/gradle-multi-project/.vscode/settings.json:17]: You can remove everything down from here as we are not testing for these settings, only `java`. You can also remove all the `typescript.*` and `javascript.*` settings above.

File: test-fixtures/gradle-multi-project/README.md [hunk 5]
```diff
@@ -0,0 +1,13 @@
+# Example Gradle Multi Project
+
+An example Gradle project used in the tests.
+
+## Formatting massive files
```
Comment (badsyntax) [inline test-fixtures/gradle-multi-project/README.md:5]: You can remove all the text from here and below as it's not relevant for this test-fixture.

**Review by badsyntax — CHANGES_REQUESTED** [review]
Added a couple of comments, minor changes. Thanks!

**Review by badsyntax — APPROVED** [review]

**Comment by badsyntax** [comment]
I decided to make the final changes to get a release out quicker. Many thanks again for your contribution.

**Comment by badsyntax** [comment]
This change is released with [1.1.4](https://github.com/badsyntax/vscode-spotless-gradle/releases/tag/1.1.4) and it seems to be working as expected on a multi-project 👍

**Comment by forestaa** [comment]
Great! Thank you for all your support!!
It seems to be working on my project.
You are so kind and I'm happy to contribute to this extension!

**Pull request merged**